mozilla/blurts-server
# Add property-based test, fix two bugs in stats calculation

# Description

So this is a big PR, but it mainly adds [property-based tests](https://fast-check.dev/docs/introduction/why-property-based/) for the `getDashboardSummary` function. I figured I'd add those after the report about the changing stats data, to rule out errors in the calculation itself.

Basically, what the test does is a bunch of runs of the following:

1. Generate some arbitrary statistics.
2. Create the breaches and scan results to match those statistics.
3. Pass those breaches and scan results to `getDashboardSummary`, then verify that the result matches the numbers generated in step 1. (This is the "property" that is verified to hold, hence the name.)

Step 1 is done randomly, and then when the test fails, it returns the seed that resulted in the invalid test result's input data. I found two bugs this way, and I've explicitly added the applicable seeds in their individual tests: 7f94e736e24d1f9202a90caf5ce9ad1abc53007f and 646a960f3199b54397c7755e9c8793d35b0ee2bc.

With those two bugs fixed, the property-based tests are now succeeding for all generated data, giving more confidence that it's actually working correctly regardless of the user's specific situation.

# How to test

Run `npm test`. Also try introducing a bug into `getDashboardSummary`, to see what the failures look like.

(And maybe try finding an account with only credit card and bank account breaches, and see how they don't get counted.)

# Checklist (Definition of Done)
- [ ] Product Owner accepted the User Story (demo of functionality completed) or waived the privilege. N/A
- [ ] Jira ticket has been updated (if needed) to match changes made during the development process. N/A
- [ ] Jira ticket has been updated (if needed) with suggestions for QA when this PR is deployed to stage. N/A
